### DISCUSSION TITLE
Serial1 and Serial2 port usage and reassignment of pins on arduino-pico

### DISCUSSION THREAD

**Discussion by kenb2054** [earlephilhower/arduino-pico#199] [Q&A] [answered]
The arduino-pico documentation states that it supports Serial1 (UART0) and Serial2 (UART1) and also allows reassignment of pins to these ports.  
I am using Arduino IDE on Windows 10 connected to arduino-pico.
Whilst I have successfully run the following example code using Serial1 using its default pins, I received error messages when trying to run the same code using Serial2.    I also had no success when trying to reassign the pin numbers to the Serial1 port.   Can you please assist in pointing out what the issue is?

    //The following example code was successful

    void setup() {
      // initialize both serial ports:
      Serial.begin(9600);
      Serial1.begin(9600);
    }

    void loop() {
      // read from port 1, send to port 0:
      if (Serial1.available()) {
        int inByte = Serial1.read();
        Serial.write(inByte);
      }

      // read from port 0, send to port 1:
      if (Serial.available()) {
        int inByte = Serial.read();
        Serial1.write(inByte);
      }
    }

When I try the same code as above but substituting Serial2 for Serial1, I get the following error message:

    Arduino: 1.8.13 (Windows 10), Board: "Raspberry Pi Pico"
    sketch\MultiSerial_Demo_Serial2.ino.cpp.o: In function `setup':

    F:\TECHNICAL\ARDUINO PICO SKETCHES\MultiSerial_Demo_Serial2/MultiSerial_Demo_Serial2.ino:8: undefined reference to 
    `_UART2_'
    sketch\MultiSerial_Demo_Serial2.ino.cpp.o: In function `loop':

    F:\TECHNICAL\ARDUINO PICO SKETCHES\MultiSerial_Demo_Serial2/MultiSerial_Demo_Serial2.ino:22: undefined reference to 
    `_UART2_'
    collect2.exe: error: ld returned 1 exit status
    exit status 1
    Error compiling for board Raspberry Pi Pico.

Also, when I try substituting pin numbers for Serial1 using the following code, I also get an error message:

    //Example to reassign pin numbers to Serial1

    void setup() {

      // reassign pin numbers for Serial1
      Serial1.setTx(16);
      Serial1.setRx(17);
  
      // initialize both serial ports:
      Serial.begin(9600);
      Serial1.begin(9600);
    }

    void loop() {
      // read from port 1, send to port 0:
      if (Serial1.available()) {
        int inByte = Serial1.read();
        Serial.write(inByte);
      }

      // read from port 0, send to port 1:
      if (Serial.available()) {
        int inByte = Serial.read();
        Serial1.write(inByte);
      }
    }

Error message when reassigning pin numbers is:

    Arduino: 1.8.13 (Windows 10), Board: "Raspberry Pi Pico"

    F:\TECHNICAL\ARDUINO PICO SKETCHES\MultiSerial_Demo_Pin_Substitute\MultiSerial_Demo_Pin_Substitute.ino: In function 
    'void setup()':

    MultiSerial_Demo_Pin_Substitute:6:11: error: 'class arduino::UART' has no member named 'setTx'
       Serial1.setTx(16);
               ^~~~~

    MultiSerial_Demo_Pin_Substitute:7:11: error: 'class arduino::UART' has no member named 'setRx'
       Serial1.setRx(17);
               ^~~~~
    exit status 1
    'class arduino::UART' has no member named 'setTx'

**Comment by earlephilhower** [comment]
You've got the wrong capitalization for the calls.  It's `setTX` and `setRX`.  You also need to make sure the selected pins are legal for the RP2040 (see the pinout diagrams from the RPI foundation).

**Reply by kenb2054** [reply]
Thanks.  I tried reassigning the pins with the correct capitalization as follows but still got the same error message:
Serial1.setTX(16);
Serial1.setRX(17);
According to the RP2040 pinout diagram, UART0 TX/RX can use GPIO pins 0/1, 12/13, 16/17 and 28/29 which I have tried to no avail.
I did a fresh install of Arduino IDE using Board Manager for Arduino Mbed OS RP2040 ver 2.1.0.
Do I need to also include any specific libraries for this to work?

**Reply by kenb2054** [reply]
Error message is same as previous but references setRX instead of incorrect syntax of setRx which I had before:

    Arduino: 1.8.15 (Windows 10), Board: "Raspberry Pi Pico"

    C:\Users\Ken\Documents\Arduino\MultiSerial_Demo_Pin_Substitute\MultiSerial_Demo_Pin_Substitute.ino: In function 'void             setup()':
    MultiSerial_Demo_Pin_Substitute:6:11: error: 'class arduino::UART' has no member named 'setTX'
   Serial1.setTX(16);
           ^~~~~

    MultiSerial_Demo_Pin_Substitute:7:11: error: 'class arduino::UART' has no member named 'setRX'
       Serial1.setRX(17);
           ^~~~~

    exit status 1
    'class arduino::UART' has no member named 'setTX'

**Reply by Andy2No** [reply]
I may have replied in the wrong box :)  See below.

**Comment by Andy2No** [comment]
For the official stm32 arduino core, you have to instantiate them, like this:

HardwareSerial Serial2(PA3, PA2);  // rx, tx
HardwareSerial Serial3(PB11, PB10);

or just

HardwareSerial Serial2;

You might need to #include <HardwareSerial.h> too, I'm not sure.

Perhaps that applies here too?

See: https://github.com/stm32duino/wiki/wiki/API#hardwareserial

**Comment by Andy2No** [comment, marked as answer]
It compiles okay for me, for the Raspberry Pi Pico,  kenb2054.  All I did was correct these two lines:

```,  // reassign pin numbers for Serial1
  Serial1.setTX(16);
  Serial1.setRX(17);
```

- Sketch uses 56904 bytes (2%) of program storage space. Maximum is 2093056 bytes.
- Global variables use 12184 bytes (4%) of dynamic memory, leaving 249960 bytes for local variables. Maximum is 262144 bytes.

Have you upgraded to the latest version?  I'm on 1.8.0.  Check your version in Tools->Board->Boards manager.

**Reply by kenb2054** [reply]
Hi,
I did a fresh install of Arduino IDE using Board Manager for Arduino Mbed OS RP2040 ver 2.1.0. and it lists the board as Raspberry Pi Pico.     The Board manager only offers an older version of 2.0.0 not 1.8.0 so not sure if I can downgrade.

**Reply by Andy2No** [reply]
Earle's core (the one this repo is about) is called Raspberry Pi Pico/RP2040, and the latest version is 1.8.0.

I compiled your example, with the correction, after selecting the Pi Pico for that core, in the boards manager, under:

Tools->Board->Raspberry Pi RP2040 Boards(1.8.0)->Raspberry Pi PIco

-after installing the core from the Boards manager:


![Pi Pico - board manager](https://user-images.githubusercontent.com/8021302/121300097-46c4b400-c8ee-11eb-8504-1e64ffeab3f5.gif)


I used Pico in the search box, at the top.   If you don't see it  (the bottom one, in this picture), you need to add the url for .json file that tells the boards manager where to find it.   It's in the Readme, on the Code tab for this project, under "Installing via Arduino Boards Manager".

**Reply by kenb2054** [reply]
Thanks very much Andy for the instruction.     That solved my problem.    Thanks Earle for the great Arduino port for pico.    I've learnt a lot very quickly.

**Reply by Andy2No** [reply]
You're welcome, Ken.

**Comment by dj1encore** [comment]
I am able to read from Serial1 using the default UART TX and RX pins GP0 and GP1. If I try to set the pins for Serial2, then Serial and Serial1 stop working. My windows 10 machine also reports an unknown USB device.

Serial2.setRX(4);
Serial2.setTX(5);

When I remove the Serial2 specific pins the system works as expected, however, I am not able to read the output from Serial2.

I have also tried the latest update fixes #472 with no success.

**Reply by earlephilhower** [reply]
Those are not legal pins.  Only certain ones can be used by the RP2040 chip itself.  See https://learn.adafruit.com/assets/99339

**Reply by dj1encore** [reply]
GP4 and GP5 are UART1 pins. Unless I am addressing the pins incorrectly.I thought that all references were taken from the GPIO reference.

**Reply by earlephilhower** [reply]
You have them backwards, I think.  RX==5, TX==4.

**Reply by dj1encore** [reply]
Sorry my bad, RX ==5 and TX ==4, this works fine, Thank you.
